src/TypeParser/JMSTypeParser.php

```diff
@@ -96,13 +98,18 @@ private function parseType(array $typeInfo, bool $isSubType = false): PropertyTy
         throw new InvalidTypeException(sprintf('Unknown JMS property found (%s)', var_export($typeInfo, true)));
     }
 
-    private function getCollectionClass(string $name): ?string
+    public function getTraversableClass(string $name): ?string
     {
         switch ($name) {
             case self::TYPE_ARRAY_COLLECTION:
                 return ArrayCollection::class;
+            case self::TYPE_GENERATOR:
+                return \Generator::class;
+            case self::TYPE_ARRAY_ITERATOR:
+            case self::TYPE_ITERATOR:
+                return \ArrayIterator::class;
             default:
-                return is_a($name, Collection::class, true) ? $name : null;
+                return is_a($name, \Traversable::class, true) ? $name : null;
```

Comment:
this might be problematic: traversable is an interface, not a class. if the callee expects to be able to instantiate what they get, we need to keep the collection.

Reply:
ah, you are correct. and i agree its not the parers job to select concrete implementations.

the Collection interface extends the iteratoraggregate interface (via ReadableCollection), and therefore using a doctrine ArrayCollection is a valid choice to satisfy the traversable interface. 

this is still a BC break (which is ok as we jump to 2.x) as the using class probably checks for the Collection interface. we should explain in the changelog that this change happens and that one should handle any traversable when using the output. (using the `is_a` check, the generator could already support this even when working with 1.x, as the collection is_a traversable)